### PR TITLE
Repeat

### PR DESCRIPTION
**Pull request by kdsudac**
Adding `repeat_limit` to PhaseOptions to enable retry loop (without falling into infinite loop).  
- Documented usage with `examples/repeat.py`.
- Includes unit tests.

### CODE REVIEW

File: test/core/exe_test.py [hunk 1]
```diff
@@ -76,3 +94,24 @@ def test_class_string(self):
...
+    self.test_state = mock.MagicMock(spec=TestState,
+        plug_manager=plugs.PlugManager(), logger=mock.MagicMock())
+    self.test_state.plug_manager.initialize_plugs([UnittestPlug])
+    self.phase_executor = PhaseExecutor(self.test_state)
+
+  def test_execute_phases(self):
+    results = []
+    for outcome in self.phase_executor.execute_phases(
+        [phase_two, phase_repeat], None):
+      results.append(outcome)
```
Comment (Kenadia) [inline test/core/exe_test.py]: nit: these four lines can be
```python
    results = list(self.phase_executor.execute_phases(
        [phase_two, phase_repeat], None))
```

File: test/core/exe_test.py [hunk 2]
```diff
@@ -76,3 +94,24 @@ def test_class_string(self):
...
+
+    self.assertEqual(5, len(results))
+    self.assertEqual(PhaseResult.CONTINUE, results[0].phase_result)
+    for i in xrange(1, 5):
+      self.assertEqual(PhaseResult.REPEAT, results[i].phase_result)
+
+
```
Comment (Kenadia) [inline test/core/exe_test.py]: nit: extra newlines

File: examples/repeat.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+  if not result:
+    print "Invalid result in phase_repeat_with_limit, will retry"
+    return openhtf.PhaseResult.REPEAT
+
+if __name__ == '__main__':
+  test = openhtf.Test(phase_repeat, phase_repeat_with_limit)
+  test.execute(test_start=lambda: 'RepeatDutID')
+
+
```
Comment (Kenadia) [inline examples/repeat.py]: nit: extra newlines

File: examples/repeat.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  def run(self):
+    """Increments counter and raises an exception for first two runs."""
+    self.count += 1
+    print 'FailTwicePlug: Run number %s' % (self.count)
+    if self.count < 3:
+      raise RuntimeError('Fails a couple times')
+
+    return True
+
+class FailAlwaysPlug(plugs.BasePlug):   # pylint: disable=no-init
```
Comment (Kenadia) [inline examples/repeat.py]: nit: unnecessary pylint comment here and above

File: examples/repeat.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+In some cases it's desirable to repeat a phase if there is a temporary error
+or an undesired result.  Repeats can be triggered by returning
+PhaseResult.REPEAT from a phase.  The number of repeats for a particular
+phase can be limited specifying a PhaseOptions.repeat_limit.
+"""
+
+import openhtf
+import openhtf.plugs as plugs
+from openhtf.util import conf
```
Comment (Kenadia) [inline examples/repeat.py]: nit: unused import

File: examples/repeat.py (hunk 3 above)
Comment (kdsudac) [inline examples/repeat.py]: Made suggested changes

File: examples/all_the_things.py [hunk 6]
```diff
@@ -50,6 +50,8 @@ def example_monitor(example):
 @htf.plug(prompts=user_input.UserInput)
 def hello_world(test, example, prompts):
   """A hello world test phase."""
+  import ipdb
+  ipdb.set_trace()
```
Comment (jettisonjoe) [inline examples/all_the_things.py]: Guessing the intention wasn't to check this `set_trace()` in?

File: examples/repeat.py (hunk 3 above)
Comment (jettisonjoe) [inline examples/repeat.py]: I still see extra newlines at the end of this file.

**Comment by kdsudac** [comment]
Removed extra blank lines... Was using +1/thumbs up as "Done".

**Comment by kdsudac** [comment]
Can someone with correct access merge this PR?

**Comment by jettisonjoe** [comment]
Need a second maintainer to review and merge. I'm the only one who has reviewed this PR so far.

**Comment by jettisonjoe** [comment]
@fahhem @grybmadsci

File: examples/with_plugs.py [hunk 7]
```diff
@@ -76,6 +68,8 @@ class PingDnsA(PingPlug):
 class PingDnsB(PingPlug):
   host = '8.8.4.4'
 
+# Note: phase name and total_time measurement use {} formatting with args
+# passed into the phase so each phase has a unique name.
 @htf.PhaseOptions(name='Ping-{pinger.host}-{count}')
 @plugs.plug(pinger=PingPlug.placeholder)
```
Comment (fahhem) [inline examples/with_plugs.py]: existing: `htf.plug(...)`

File: openhtf/__init__.py [hunk 8]
```diff
@@ -311,6 +312,8 @@ class PhaseOptions(mutablerecords.Record('PhaseOptions', [], {
         otherwise only the TestApi will be passed in.  This is useful if a
         phase needs to wrap another phase for some reason, as
         PhaseDescriptors can only be invoked with a TestState instance.
+    repeat_limit:  Maximum number of repeats.  None indicates a phase will
+        be repeated infinitely as long as PhaseResult.REPEAT returned.
```
Comment (fahhem) [inline openhtf/__init__.py]: is returned*

File: openhtf/core/phase_executor.py [hunk 9]
```diff
@@ -177,13 +177,14 @@ def execute_phases(self, phases, teardown_func):
 
     Args:
       phases: List of phases to execute.
-      teardown_func: 
+      teardown_func:
```
Comment (fahhem) [inline openhtf/core/phase_executor.py]: teardown_func: Function or phase to execute at the end of all phases regardless of errors

File: test/core/exe_test.py [hunk 10]
```diff
@@ -20,11 +20,18 @@
 import openhtf
 from openhtf import core
 from openhtf import plugs
+from openhtf import PhaseResult
```
Comment (fahhem) [inline test/core/exe_test.py]: Other maintainers: Should the tests do the same `import openhtf as htf` as the examples? I think they should

File: openhtf/core/phase_executor.py (hunk 9 above)
Comment (kdsudac) [inline openhtf/core/phase_executor.py]: Done

**Comment by fahhem** [comment]
LGTM

**Pull request merged**